api/app/Http/Controllers/Forms/PublicFormController.php

```diff
@@ -90,10 +90,38 @@ public function answer(AnswerFormRequest $request, FormSubmissionProcessor $form
         $form = $request->form;
         $isFirstSubmission = ($form->submissions_count === 0);
         $submissionId = false;
+        $submissionHash = $request->get('submission_hash') ?? null;
+        if ($submissionHash) {
+            $submissionHash = Hashids::decode($submissionHash);
+            $submissionId = (int)($submissionHash[0] ?? null);
+        }
+
+        $isPartial = $request->get('is_partial') ?? false;
+        if ($isPartial) {
+            $submissionResponse = $form->submissions()->updateOrCreate([
+                'id' => $submissionId
+            ], [
+                'data' => $request->all(),
+                'status' => FormSubmission::STATUS_PARTIAL
+            ]);
+            $submissionId = $submissionResponse->id;
+
+            return $this->success([
+                'message' => 'Partial submission saved',
+                'submission_hash' => Hashids::encode($submissionId)
+            ]);
+        }
 
         $submissionData = $request->validated();
         $completionTime = $request->get('completion_time') ?? null;
-        unset($submissionData['completion_time']); // Remove completion_time from the main data array
+        // Remove extra fields from the main data array
+        unset($submissionData['completion_time']);
+        unset($submissionData['submission_hash']);
+
+        // Add submission_id to the submission data if it exists
+        if ($submissionId) {
+            $submissionData['submission_id'] = $submissionId;
+        }
 
         $job = new StoreFormSubmissionJob($form, $submissionData, $completionTime);
 
```

Comment:
Below, instead of 

```
else {
            StoreFormSubmissionJob::dispatch($form, $submissionData, $completionTime);
        }
```

we could simply do `$job->dispatch()` or similar to better reafactor